Add coccinelle support

clumens opened the pull request:
This splits a single patch out of #1684.  Coccinelle is kind of hard to use, but I think it's going to be beneficial for refactorings and patch review.  This is the same as poki's patch, except with a couple points:

* I have dropped the test case portion.  I'm not sure I want to maintain a test suite for this particular thing.
* I have addressed the other comments on that original patch, except for modifying the SPDX text.  I'm not sure what the legalities are around modifying license text that someone else added.  I thought it best to discuss that separately first.
* I have added my own .cocci file that checks for where pcmk__str_empty could be used, and then added another patch that addresses what it found.

And then there's a couple other things that could be done:

* This could potentially be run out of the .travis.yml file, but then we'd also need the builders to have coccinelle installed on them.
* I would like the ref-passed-variables-inited.cocci to not do anything, which means we need to apply it and fix up everything that it finds.  I figure that's worth a discussion first.
* There will be a ton of other things this can help with in the future, for example my pcmk__strcmp patch set.  I'm hoping to get this in and then use it to aid in that refactoring.

kgaillot commented:
I love the capability, hopefully it will be worth its complexity. Suggestions:

* Pull in poki's commit as-is, and make your corrections to it in a separate commit.
* Move this to a new devel/ subdirectory instead of maint/. I'm thinking of reserving maint/ for stuff related to releases, CI, etc., whereas devel/ would have helpers usable by any Pacemaker developer even if they aren't a project maintainer.
* I'm not sure about dropping poki's tests since I can imagine changing that recipe will be a nail-biting experience without it. That doesn't mean we need tests for every recipe, but they might come in handy for ones that are more complicated and/or likely to be changed in the future.
* SPDX is not part of the license, just a way to declare it, so we can change that as desired.
* I'll email poki about potentially relicensing the FSFAP stuff under GPL so we don't need yet another license with the project.

clumens commented:
>     * Pull in poki's commit as-is, and make your corrections to it in a separate commit.

I should have stuck with my instincts and done it that way to begin with.

>     * Move this to a new devel/ subdirectory instead of maint/. I'm thinking of reserving maint/ for stuff related to releases, CI, etc., whereas devel/ would have helpers usable by any Pacemaker developer even if they aren't a project maintainer.

A related question that I forgot to ask the first time around - this should probably have a Makefile.am so it gets distributed as part of the source tarball, right?

>     * I'm not sure about dropping poki's tests since I can imagine changing that recipe will be a nail-biting experience without it. That doesn't mean we need tests for every recipe, but they might come in handy for ones that are more complicated and/or likely to be changed in the future.

Okay, I can add that back in.  I'm not sure how to run them.  We don't want it to be part of the top-level "make check" because that will require coccinelle.  But I do kind of want it to happen from some automated place.

>     * SPDX is not part of the license, just a way to declare it, so we can change that as desired.
> 
>     * I'll email poki about potentially relicensing the FSFAP stuff under GPL so we don't need yet another license with the project.

Okay, sounds good.

kgaillot commented:
> A related question that I forgot to ask the first time around - this should probably have a Makefile.am so it gets distributed as part of the source tarball, right?

Yes, similar to what maint has, I think these would be EXTRA_*


> Okay, I can add that back in. I'm not sure how to run them. We don't want it to be part of the top-level "make check" because that will require coccinelle. But I do kind of want it to happen from some automated place.

Agreed that it shouldn't be done by default. A make target in maint is fine.


From poki:

>> On 7/27/20 11:25 PM, Ken Gaillot wrote:
>>> I have a Pacemaker question for you ... how would you feel about us
>>> relicensing your coccinelle/gdbhelpers stuff under GPLv2 instead of
>>> FSFAP? I'm thinking I'd rather avoid adding another license to the mix
>>> (we already have 4).
>> 
>> Signed-off-by: Jan Pokorný <poki@poki.me>
>> 
>> Let me know if you want some ack at some PR.
>
> Reusing already applied CC license would be even better, though, to
> denote beyond-code ... it's not needed for udage, just as documentation
> once one already knows.

So either GPLv2 or CC-BY-SA-4.0 would be fine for those. Let's pull in his gdbhelpers commits as well and move those to devel/ too.

clumens commented:
> Let's pull in his gdbhelpers commits as well and move those to devel/ too.

Sounds good.  I think this would work best as a followup PR, just because this one has additional coccinelle-related stuff (my pcmk__str_empty patches).

clumens commented:
I think that's everything except applying the fixes that ref-passed-variables-inited.cocci finds.

kgaillot commented on GNUmakefile:
```diff
@@ -437,7 +437,7 @@ clang:
 # Run the "make cocci" target to just output what would be done, or "make cocci-inplace"
 # to apply the changes to the source tree.
-COCCI_FILES ?= maint/coccinelle/*.cocci
+COCCI_FILES ?= devel/coccinelle/*.cocci
```
I would move all the cocci stuff to devel/Makefile.am

clumens replied:
And then we run `make -C devel cocci`?

kgaillot replied:
exactly

clumens commented:
I'm going to need to think about this one a little bit longer.  ref-passed-variables-inited.cocci looks like it's adding FIXMEs to GHashTableIter variables that do get properly initialized.

clumens commented:
Rebased, made it so COCCI_FILES is only a subset of existing files (ref-passed-variables-inited.cocci is disabled for the moment), and removed the WIP.
